### PR TITLE
Removes ldap_sort while preserving sorting

### LINKED ISSUE #3: Remove the call to ldap_sort from Zend\Ldap\Ldap::search
Currently ```Zend\Ldap\Ldap::search()``` internally calls ```ldap_sort``` which is already deprecated in the OpenLDAP-libraries and will become deprecated in PHP. As the sorting only is done client-sided and we can not be sure that there might be other data on the server that should be in between the sorting is not really meaningfull. 

We should look for a way to sort the data with the "default" PHP-sorting-functions

### PR DESCRIPTION
**Pull request by heiglandreas**
This PR removes the call to the now deprecated ldap_sort-function wile
still preserving the sort-functionality.

For that the DefaultIterator is replaced with a SortableIterator (which
extends the DefaultIterator) that uses natural sorting (natcasesort).

This fixes #3

### CODE REVIEW

File: src/Collection/SortableIterator.php [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+ * THE SOFTWARE.
+ *
+ * @category
+ * @author    Andreas Heigl<andreas@heigl.org>
+ * @copyright ©2015-2015 Andreas Heigl
+ * @license   http://www.opesource.org/licenses/mit-license.php MIT-License
+ * @version   0.0
+ * @since     22.07.15
+ * @link      https://github.com/heiglandreas/
+ */
```
Comment (heiglandreas) [inline src/Collection/SortableIterator.php]: ZF Header please!

File: src/Collection/SortableIterator.php [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+namespace Zend\Ldap\Collection;
+
+use Zend\Ldap\Ldap;
+
+class SortableIterator extends DefaultIterator
+{
+    /**
+     * @var array $entries
```
Comment (heiglandreas) [inline src/Collection/SortableIterator.php]: Can we get a more detailed description on what's the purpose of that property?

File: src/Collection/SortableIterator.php [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+     */
+    protected $entries = array();
+
+    public function __construct(Ldap $ldap, $resultId)
```
Comment (heiglandreas) [inline src/Collection/SortableIterator.php]: Please add a description that also contains the parameter-description in here

File: src/Collection/SortableIterator.php [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    {
+        parent::__construct($ldap, $resultId);
+
+        $identifier = ldap_first_entry(
+            $ldap->getResource(),
+            $resultId
+        );
+
+        $identifier;
```
Comment (heiglandreas) [inline src/Collection/SortableIterator.php]: this looks somewhat displaced

File: src/Collection/SortableIterator.php [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+            );
+        }
+
+
+    }
+
+    /**
+     * Sort the iterator by the given attribute
+     *
+     * @param string $sortAttribute
```
Comment (heiglandreas) [inline src/Collection/SortableIterator.php]: Can we please have a parameter-description here?

File: src/Collection/SortableIterator.php [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+            );
+
+            if (isset($attributes[$sortAttribute][0])) {
+                $this->entries[$key]['sortValue'] = $attributes[$sortAttribute][0];
+            }
+        }
+
+        return usort($this->entries, function($a,$b){
+            return strnatcasecmp($a['sortValue'], $b['sortValue']);
+        });
```
Comment (heiglandreas) [inline src/Collection/SortableIterator.php]: Shall we inject the sorting-function as second parameter? 

Would change the function-footprint to ```public function sort($sortAtribute, callable $sortFunction = null)``` and if ```$sortFunction``` is null would default to the one implemented now.

Any ideas @weierophinney @Maks3w ?

File: src/Collection/SortableIterator.php (hunk 1 above)
Comment (Maks3w) [inline src/Collection/SortableIterator.php]: :)

File: src/Collection/SortableIterator.php [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+     *
+     * @return void
+     */
+    public function rewind()
+    {
+        reset($this->entries);
+        $nextEntry = current($this->entries);
+        $this->current = $nextEntry['resource'];
+    }
+
```
Comment (heiglandreas) [inline src/Collection/SortableIterator.php]: to be removed

File: src/Collection/SortableIterator.php (hunk 1 above)
Comment (heiglandreas) [inline src/Collection/SortableIterator.php]: Yeah, Still not configured my IDE to use proper templates for different projects..... :wink:

File: src/Collection/SortableIterator.php [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+    }
+
+    /**
+     * Sort the iterator by the given attribute
+     *
+     * @param string $sortAttribute
+     *
+     * @return self
```
Comment (Maks3w) [inline src/Collection/SortableIterator.php]: returns boolean

**Comment by heiglandreas** [comment]
Should this not be included in the next 2.5-release as it fixes failing tests in PHP7?

**Comment by Maks3w** [comment]
No because PHP7 is allowed_failure. Also this adds a new Iterator. In a few moments I'll give you feedback about the Iterator

File: src/Ldap.php [hunk 9]
```diff
@@ -923,17 +923,16 @@ public function search($filter, $basedn = null, $scope = self::SEARCH_SCOPE_SUB,
...
+
+        $iterator = new Collection\SortableIterator($this, $search);
+
         if ($sort !== null && is_string($sort)) {
-            ErrorHandler::start(E_WARNING);
-            $isSorted = ldap_sort($resource, $search, $sort);
-            ErrorHandler::stop();
+            $isSorted = $iterator->sort($sort);
             if ($isSorted === false) {
                 throw new Exception\LdapException($this, 'sorting: ' . $sort);
```
Comment (Maks3w) [inline src/Ldap.php]: Why is this exception required?

If required I think can be placed in sort() method

**Comment by Maks3w** [comment]
I like the idea of put the sort logic in the Collection namespace.

Two things:

1. I think the DefaultIterator could be host the logic.
2. Create a unit test for the new class.

File: src/Collection/SortableIterator.php (hunk 6 above)
Comment (Maks3w) [inline src/Collection/SortableIterator.php]: I like the idea of inject a custom sort logic.

But if I insert my own logic then `$sortAttribute` could be optional? If yes then I think should be two different methods.

**Comment by heiglandreas** [comment]
I was thinking about adding the sort-logic to the DefaultIterator as well as it would eliminate the need for a separate SortingIterator. Especially as it's a hard-coded dependency so using the SortingIterator would defacto eliminate the need for the DefaultIterator. On the other hand I wanted to leave the existing code intact as much as possible. 

But I think the pros for adding the sort-functionality to the DefaultIterator outweigh the cons. Therefore I will add the logic to the DefaultIterator (and add tests accordingly)

File: src/Ldap.php (hunk 9 above)
Comment (heiglandreas) [inline src/Ldap.php]: It's an artifact of the current implementation. If I move it to the sort() method (which in my eyes makes much more sense) I would change the return value of the sort method from ```boolean``` to ```self``` (or whatever is appropriate in that file).

Comment (Maks3w) [inline src/Ldap.php]: Please move into sort method. The return could be just `void`

**Comment by Maks3w** [comment]
@heiglandreas Please don't rewrite the commit while the review is in progress. I lost the progress of changes and every time I suppose the code must be full reviewed again.

File: src/Collection/SortableIterator.php [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+    protected $entries = array();
+
+    /**
+     * Create an instance of the SortableIterator.
+     *
+     * @param Ldap     $ldap     The LDAP-Object currently used
+     * @param resource $resultId A reference to the resultset returned by ldap_search
+     *
+     * @throws \Zend\Ldap\Exception\LdapException
+     * @return void
```
Comment (Maks3w) [inline src/Collection/SortableIterator.php]: remove return for constructors

**Comment by heiglandreas** [comment]
Sorry @Maks3w! Will stop it and try to remember for the future!

**Comment by Maks3w** [comment]
Remember to execute php-cs-fixer before push :)

File: src/Collection/SortableIterator.php (hunk 6 above)
Comment (heiglandreas) [inline src/Collection/SortableIterator.php]: ```sortAttribute``` is required in any case as the value of the first entry of that sort-attribute is passed as sort-parameter. It's the only thing that is available at that point as the complete dataset is not yet retrieved.

**Comment by heiglandreas** [comment]
@Maks3w I've reworked the sorting.

File: test/SortTest.php [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+        );
+
+        $iterator = new DefaultIterator($this->getLdap(), $search);
+        $sortFunction = function ($a, $b) use ($lSorted) {
+            if (array_search($a, $lSorted) % 2 === 0) {
+                return -1;
+            }
+
+            return 1;
+        };
```
Comment (heiglandreas) [inline test/SortTest.php]: This sort function has to be optimized to work with PHP7 or HHVM

**Comment by mwillbanks** [comment]
Was merged into develop but pointed at master, closing PR as it is merged.

**Comment by jkrehm** [comment]
Any guess when this will get into `master` and be part of an official release? Right now I have my company's project pointed at the `develop` branch specifically for this issue (we're taking care of PHP 7 issues in our application so clients can start using it sometime next year), but I'd of course love to switch back to a stable release version once the work done here makes it to master.

**Comment by heiglandreas** [comment]
I'm trying to get that merged this year.

**Comment by jkrehm** [comment]
Awesome, thanks for both your work on the library and your persistence with moving this change forward.

**Pull request closed without merging**